### ISSUE TITLE
New-SymLink does not appear to work, and is missing Usage and Syntax guidance

### ISSUE THREAD

**Issue by bitbasher** [Pscx/Pscx#102]
i ran into some problems getting any command to work, but i did overcome them

but the basic New-SYmLink command does not appear to work .. and worse, it fails silently

doing this

> New-Symlink ./parts_library D:\repositories\FreeCAD-library\

or this

>  New-Symlink  C:\Users\Jeff\AppData\Roaming\FreeCAD\Mod\parts_library D:\repositories\FreeCAD-library\

return no error message .. but also do NOT create the symbolic link.

in the absence of syntax guidance i even tried this 

`New-Symlink D:\repositories\FreeCAD-library\ C:\Users\Jeff\AppData\Roaming\FreeCAD\Mod\parts_library
`
which gave this response

> New-Symlink : New-Symlink's TargetPath parameter has an invalid path of
> 'C:\Users\Jeff\AppData\Roaming\FreeCAD\Mod\parts_library': The path
> 'C:\Users\Jeff\AppData\Roaming\FreeCAD\Mod\parts_library' supplied for TargetPath must exist.

So .. i think that it has to be "new-symlink newLink d:\targetlink".

Reading the code i see that it detects that the target is a directory so i don't need to do '/d' as was needed on the old "mklink" command

I have created a couple of Pull requests with changes to the README to document the things i had to do to enable the commands in WIndows 11

**Comment by hrxn** [comment]
FWIW, you can use this directly with PowerShell

```
PS C:\> [System.IO.File]::CreateSymbolicLink

OverloadDefinitions
-------------------
static System.IO.FileSystemInfo CreateSymbolicLink(string path, string pathToTarget)


PS C:\>
```

Should work with PowerShell 7.2 (or newer), could you try that?

**Comment by bitbasher** [comment]
sorry .. i missed your reply .. i will try this and get back to you

**Comment by bitbasher** [comment]
yeah .. that did not work
`Windows PowerShell
Copyright (C) Microsoft Corporation. All rights reserved.

Install the latest PowerShell for new features and improvements! https://aka.ms/PSWindows

PS C:\Users\Jeff> [System.IO.File]::CreateSymbolicLink
PS C:\Users\Jeff> cd .\AppData\Roaming\FreeCAD\Mod\
PS C:\Users\Jeff\AppData\Roaming\FreeCAD\Mod> ls


    Directory: C:\Users\Jeff\AppData\Roaming\FreeCAD\Mod


Mode                 LastWriteTime         Length Name
----                 -------------         ------ ----
d-----         12/6/2024   1:08 AM                3D_Printing_Tools
d-----        12/27/2024   2:48 PM                addFC
d-----        11/25/2024   6:17 PM                Assembly4
d-----        11/28/2024  12:37 PM                boltsfc
d-----        11/28/2024  12:29 PM                btl
d-----        11/25/2024   5:48 PM                Curves
d-----        11/25/2024   6:17 PM                Defeaturing
d-----        11/25/2024   6:17 PM                DynamicData
d-----         12/5/2024  11:01 PM                fasteners
d-----        11/28/2024  12:46 PM                FrameForge
d-----        12/12/2024   8:39 PM                freecad.gears
d-----        11/25/2024   8:40 PM                FreeGrid
d-----        11/25/2024   5:49 PM                Gridfinity
d-----        12/12/2024   8:39 PM                InventorLoader
d-----        11/28/2024   2:13 PM                lattice2
d-----        11/25/2024   5:50 PM                LCInterlocking
d-----        12/12/2024   8:40 PM                MakerWorkbench
d-----        11/28/2024   2:13 PM                Manipulator
d-----         12/5/2024  11:02 PM                MeshRemodel
d-----         12/5/2024  11:37 PM                parts_library
d-----        11/25/2024   6:17 PM                Pyramids-and-Polyhedrons
d-----        12/12/2024   8:40 PM                sheetmetal
d-----        11/25/2024   6:17 PM                slic3r-tools
d-----        11/25/2024   6:11 PM                symbols_library
d-----        11/25/2024   6:17 PM                ThreadProfile
d-----        11/25/2024   6:12 PM                toSketch
d-----        12/12/2024   8:40 PM                woodworking
d-----        11/25/2024   6:17 PM                workfeature
d-----        11/25/2024   6:17 PM                yaml-workspace


PS C:\Users\Jeff\AppData\Roaming\FreeCAD\Mod> ls


    Directory: C:\Users\Jeff\AppData\Roaming\FreeCAD\Mod


Mode                 LastWriteTime         Length Name
----                 -------------         ------ ----
d-----         12/6/2024   1:08 AM                3D_Printing_Tools
d-----        12/27/2024   2:48 PM                addFC
d-----        11/25/2024   6:17 PM                Assembly4
d-----        11/28/2024  12:37 PM                boltsfc
d-----        11/28/2024  12:29 PM                btl
d-----        11/25/2024   5:48 PM                Curves
d-----        11/25/2024   6:17 PM                Defeaturing
d-----        11/25/2024   6:17 PM                DynamicData
d-----         12/5/2024  11:01 PM                fasteners
d-----        11/28/2024  12:46 PM                FrameForge
d-----        12/12/2024   8:39 PM                freecad.gears
d-----        11/25/2024   8:40 PM                FreeGrid
d-----        11/25/2024   5:49 PM                Gridfinity
d-----        12/12/2024   8:39 PM                InventorLoader
d-----        11/28/2024   2:13 PM                lattice2
d-----        11/25/2024   5:50 PM                LCInterlocking
d-----        12/12/2024   8:40 PM                MakerWorkbench
d-----        11/28/2024   2:13 PM                Manipulator
d-----         12/5/2024  11:02 PM                MeshRemodel
d-----        11/25/2024   6:17 PM                Pyramids-and-Polyhedrons
d-----        12/12/2024   8:40 PM                sheetmetal
d-----        11/25/2024   6:17 PM                slic3r-tools
d-----        11/25/2024   6:11 PM                symbols_library
d-----        11/25/2024   6:17 PM                ThreadProfile
d-----        11/25/2024   6:12 PM                toSketch
d-----        12/12/2024   8:40 PM                woodworking
d-----        11/25/2024   6:17 PM                workfeature
d-----        11/25/2024   6:17 PM                yaml-workspace


PS C:\Users\Jeff\AppData\Roaming\FreeCAD\Mod> [System.IO.File]::CreateSymbolicLink parts_library D:\repositories\parts_library\
At line:1 char:38
+ [System.IO.File]::CreateSymbolicLink parts_library D:\repositories\pa ...
+                                      ~~~~~~~~~~~~~
Unexpected token 'parts_library' in expression or statement.
    + CategoryInfo          : ParserError: (:) [], ParentContainsErrorRecordException
    + FullyQualifiedErrorId : UnexpectedToken

PS C:\Users\Jeff\AppData\Roaming\FreeCAD\Mod> [System.IO.File]::CreateSymbolicLink D:\repositories\parts_library\ parts_library
At line:1 char:38
+ ... m.IO.File]::CreateSymbolicLink D:\repositories\parts_library\ parts_l ...
+                                    ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
Unexpected token 'D:\repositories\parts_library\' in expression or statement.
    + CategoryInfo          : ParserError: (:) [], ParentContainsErrorRecordException
    + FullyQualifiedErrorId : UnexpectedToken

PS C:\Users\Jeff\AppData\Roaming\FreeCAD\Mod> [System.IO.File]::CreateSymbolicLink
PS C:\Users\Jeff\AppData\Roaming\FreeCAD\Mod> ls


    Directory: C:\Users\Jeff\AppData\Roaming\FreeCAD\Mod


Mode                 LastWriteTime         Length Name
----                 -------------         ------ ----
d-----         12/6/2024   1:08 AM                3D_Printing_Tools
d-----        12/27/2024   2:48 PM                addFC
d-----        11/25/2024   6:17 PM                Assembly4
d-----        11/28/2024  12:37 PM                boltsfc
d-----        11/28/2024  12:29 PM                btl
d-----        11/25/2024   5:48 PM                Curves
d-----        11/25/2024   6:17 PM                Defeaturing
d-----        11/25/2024   6:17 PM                DynamicData
d-----         12/5/2024  11:01 PM                fasteners
d-----        11/28/2024  12:46 PM                FrameForge
d-----        12/12/2024   8:39 PM                freecad.gears
d-----        11/25/2024   8:40 PM                FreeGrid
d-----        11/25/2024   5:49 PM                Gridfinity
d-----        12/12/2024   8:39 PM                InventorLoader
d-----        11/28/2024   2:13 PM                lattice2
d-----        11/25/2024   5:50 PM                LCInterlocking
d-----        12/12/2024   8:40 PM                MakerWorkbench
d-----        11/28/2024   2:13 PM                Manipulator
d-----         12/5/2024  11:02 PM                MeshRemodel
d-----        11/25/2024   6:17 PM                Pyramids-and-Polyhedrons
d-----        12/12/2024   8:40 PM                sheetmetal
d-----        11/25/2024   6:17 PM                slic3r-tools
d-----        11/25/2024   6:11 PM                symbols_library
d-----        11/25/2024   6:17 PM                ThreadProfile
d-----        11/25/2024   6:12 PM                toSketch
d-----        12/12/2024   8:40 PM                woodworking
d-----        11/25/2024   6:17 PM                workfeature
d-----        11/25/2024   6:17 PM                yaml-workspace


PS C:\Users\Jeff\AppData\Roaming\FreeCAD\Mod> [System.IO.File]::CreateSymbolicLink --help
At line:1 char:40
+ [System.IO.File]::CreateSymbolicLink --help
+                                        ~~~~
Unexpected token 'help' in expression or statement.
    + CategoryInfo          : ParserError: (:) [], ParentContainsErrorRecordException
    + FullyQualifiedErrorId : UnexpectedToken

PS C:\Users\Jeff\AppData\Roaming\FreeCAD\Mod> cmd /c mklink /d ./parts_library D:\repositories\parts_library\
Invalid switch - "parts_library".
PS C:\Users\Jeff\AppData\Roaming\FreeCAD\Mod> cmd /c mklink /d "./parts_library" "D:\repositories\parts_library\"
Invalid switch - "parts_library".
PS C:\Users\Jeff\AppData\Roaming\FreeCAD\Mod>
PS C:\Users\Jeff\AppData\Roaming\FreeCAD\Mod>
PS C:\Users\Jeff\AppData\Roaming\FreeCAD\Mod> New-Item -ItemType SymbolicLink -Path .\parts_library -Target D:\repositories\parts_library\
New-Item : Administrator privilege required for this operation.
At line:1 char:1
+ New-Item -ItemType SymbolicLink -Path .\parts_library -Target D:\repo ...
+ ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    + CategoryInfo          : PermissionDenied: (D:\repositories\parts_library\:String) [New-Item],
   UnauthorizedAccessException
    + FullyQualifiedErrorId : NewItemSymbolicLinkElevationRequired,Microsoft.PowerShell.Commands.New
   ItemCommand

PS C:\Users\Jeff\AppData\Roaming\FreeCAD\Mod>
PS C:\Users\Jeff\AppData\Roaming\FreeCAD\Mod>
PS C:\Users\Jeff\AppData\Roaming\FreeCAD\Mod>
PS C:\Users\Jeff\AppData\Roaming\FreeCAD\Mod>`

**Comment by bitbasher** [comment]
so that was messy but .. i had installed the parts_library folder using github in another location and moved it into the mods folder .. so that worked for me

i moved the folder over to my D: drive to test symlinking in power shell .. your suggested command does not work at all .. more googling showed me lots about the mklink command .. i tried that again, also failed

more googling found me the New-Item command and i made that work .. but i had to start powershell as administrator to get enough rights 
PS C:\Users\Jeff\AppData\Roaming\FreeCAD\Mod>
PS C:\Users\Jeff\AppData\Roaming\FreeCAD\Mod> New-Item -ItemType SymbolicLink -Path .\parts_library -Target D:\repositories\parts_library\


    Directory: C:\Users\Jeff\AppData\Roaming\FreeCAD\Mod


Mode                 LastWriteTime         Length Name
----                 -------------         ------ ----
d----l         2/16/2025   5:11 PM                parts_library


PS C:\Users\Jeff\AppData\Roaming\FreeCAD\Mod> ls


    Directory: C:\Users\Jeff\AppData\Roaming\FreeCAD\Mod


Mode                 LastWriteTime         Length Name
----                 -------------         ------ ----
d-----         12/6/2024   1:08 AM                3D_Printing_Tools
d-----        12/27/2024   2:48 PM                addFC
d-----        11/25/2024   6:17 PM                Assembly4
d-----        11/28/2024  12:37 PM                boltsfc
d-----        11/28/2024  12:29 PM                btl
d-----        11/25/2024   5:48 PM                Curves
d-----        11/25/2024   6:17 PM                Defeaturing
d-----        11/25/2024   6:17 PM                DynamicData
d-----         12/5/2024  11:01 PM                fasteners
d-----        11/28/2024  12:46 PM                FrameForge
d-----        12/12/2024   8:39 PM                freecad.gears
d-----        11/25/2024   8:40 PM                FreeGrid
d-----        11/25/2024   5:49 PM                Gridfinity
d-----        12/12/2024   8:39 PM                InventorLoader
d-----        11/28/2024   2:13 PM                lattice2
d-----        11/25/2024   5:50 PM                LCInterlocking
d-----        12/12/2024   8:40 PM                MakerWorkbench
d-----        11/28/2024   2:13 PM                Manipulator
d-----         12/5/2024  11:02 PM                MeshRemodel
d----l         2/16/2025   5:11 PM                parts_library
d-----        11/25/2024   6:17 PM                Pyramids-and-Polyhedrons
d-----        12/12/2024   8:40 PM                sheetmetal
d-----        11/25/2024   6:17 PM                slic3r-tools
d-----        11/25/2024   6:11 PM                symbols_library
d-----        11/25/2024   6:17 PM                ThreadProfile
d-----        11/25/2024   6:12 PM                toSketch
d-----        12/12/2024   8:40 PM                woodworking
d-----        11/25/2024   6:17 PM                workfeature
d-----        11/25/2024   6:17 PM                yaml-workspace


PS C:\Users\Jeff\AppData\Roaming\FreeCAD\Mod>

and as you can see this worked

>

**Comment by bitbasher** [comment]
hummm .. digging around for the version info for powershell revealed that i have at least two instances of it .. Windows Powershell V1 in /System32 and powershell V7 in C:\Program Files\P:owershell\7 .. maybe i was using the wrong PS?

**Comment by bitbasher** [comment]
yep .. i was using the wrong powershell .. using PS 7.4.6 i get this

> PS C:\Users\Jeff> cd .\AppData\Roaming\FreeCAD\Mod\
PS C:\Users\Jeff\AppData\Roaming\FreeCAD\Mod> cmd /c mklink /d .\parts_library 'D:\repositories\parts_library\'
symbolic link created for .\parts_library <<===>> D:\repositories\parts_library\
PS C:\Users\Jeff\AppData\Roaming\FreeCAD\Mod>


so i am happy

**Comment by oising** [comment]
@bitbasher You're confusing dotnet method() and powershell function syntax:

```powershell
[System.IO.File]::CreateSymbolicLink("parts_library", "D:\repositories\parts_library")
```

This is what you should be doing. Dotnet stuff is invoked with parentheses and comma-separated arguments. Powershell functions -- like many other shells -- use spaces and don't require parentheses

**Comment by bitbasher** [comment]
i agree .. i have been using shells on Unix, Linus, Mac OS, and Windows for long enuf to know that .. i was looking for the Power Shell equivalent to "ln -s ./symlink-to-some.thing /anotherplace/some.thing" 

I have never even seen the dotnet method you refer to so for me there is too steep a learning curve for me to go off in that direction. Thanks for the tip  tho

FYI .. i was also able to symlink using NewItem -type SymbolicLink .. blah blah so that is a workable alternative .. but i would make a shell macro to give myself the "ln -s ..." look-alike

sooner or later i will find out how to use Bash or Zsh in a windows shell .. but not today

**Comment by bitbasher** [comment]
please note .. nothing in this thread deals with the original bug .. that the NewSymlink command documented as existing in powerShell does not work and fails silently .. but i will let someone else notice this problem
